### PR TITLE
[mqtt] Support multiple, comma-separated values for on/off param of switch/contact channels

### PR DESCRIPTION
**Pull request by florian-h05**
This allows to map multiple string values to the on or the off state.

### CODE REVIEW

**Comment by lsiepel** [comment]
Would it make sense to make use of https://github.com/openhab/openhab-core/pull/4330 look at the related HomeKit PR for example configuration?

**Comment by florian-h05** [comment]
Hmm I think the core PR is related to extending the syntax of the .items file to allow definint these lists in item metadata.
This PR is about defining these lists not at Item level, but instead at Thing (binding) level. 

I would like to map the contact channel's OPEN to multiple states, this is only possible at a Thing level. The other option would be to have a string channel and transform its value at channel link level, but this is more complicated to set up.

**Comment by ccutrer** [comment]
Things (and specifically channels) already support multiple values for their config in thing files (https://github.com/ccutrer/openhab-core/blob/c3fdd86428939a0a5b01816cfbfd84ec0a86b6df/bundles/org.openhab.core.model.thing/src/org/openhab/core/model/thing/Thing.xtext#L67) - it's actually what I based the Item metadata PR on. That said, the binding still needs to handle a List for it. And the UI definitely does _not_  support a friendly way to enter native lists. I'd come asking you, @florian-h05, about that! I'd still much prefer if we could do it that way, though. While it may seem unusual, I could see a comma being a legitimate value (perhaps it's `100,0` (as in 100%, with a European-style decimal point). A pipe (`|`) seems less likely, but if we can get some nice UI support for lists, then we wouldn't have to roll the dice that no one uses that in a legitimate value.

**Comment by jimtng** [comment]
@florian-h05 I just had to change the config parameter type from `String` to `List<String>` in java, and added `multiple="true"` in the channel config xml. Then it will accept an array / list in .things file, and in the UI. See this, those two are the only things I changed, because the implementation class already supports 
 Lists (which may not be the case here). https://github.com/openhab/openhab-addons/pull/17288/files#diff-1a86ba1ab3ca46c5eb34575ba95d85c0b87453b4bfd361711829b783c557cf44

**Comment by florian-h05** [comment]
@jimtng Thanks for the hint.
I have refactored this PR to use this, would be great if you and @ccutrer could have a look at it.

File: bundles/org.openhab.binding.mqtt.generic/src/main/java/org/openhab/binding/mqtt/generic/values/OpenCloseValue.java [hunk 1]
```diff
@@ -22,34 +26,60 @@
...
  *
  * @author David Graeff - Initial contribution
  */
 @NonNullByDefault
 public class OpenCloseValue extends Value {
-    private final String openString;
-    private final String closeString;
+    private final Set<String> openStates;
+    private final Set<String> closedStates;
+    private final String openCommand;
```
Comment (ccutrer) [inline bundles/org.openhab.binding.mqtt.generic/src/main/java/org/openhab/binding/mqtt/generic/values/OpenCloseValue.java:37]: Is it possible to somehow send a command to an OpenCloseValue? You didn't introduce that in this PR, but you definitely extend it. Assuming it's not possible, I'd rather just remove it (possibly throwing an exception in getMQTTpublishValue and parseCommand, and renaming the current parseCommand to parseMessage - though if we go that way, you need to make sure to maintain the UnDefType condition from the parent class)

File: bundles/org.openhab.binding.mqtt.generic/src/main/java/org/openhab/binding/mqtt/generic/values/ValueFactory.java [hunk 2]
```diff
@@ -72,10 +76,12 @@ public static Value createValueState(ChannelConfig config, String channelTypeID)
...
+                value = new OnOffValue(config.onState == null && on != null ? List.of(on) : config.onState,
+                        config.offState == null && off != null ? List.of(off) : config.offState);
                 break;
             case MqttBindingConstants.CONTACT:
-                value = new OpenCloseValue(config.on, config.off);
+                value = new OpenCloseValue(config.onState == null && on != null ? List.of(on) : config.onState,
+                        config.offState == null && off != null ? List.of(off) : config.offState);
                 break;
             case MqttBindingConstants.ROLLERSHUTTER:
                 value = new RollershutterValue(config.on, config.off, config.stop, config.onState, config.offState,
```
Comment (ccutrer) [inline bundles/org.openhab.binding.mqtt.generic/src/main/java/org/openhab/binding/mqtt/generic/values/ValueFactory.java:87]: do you not also need to handle multiple values for Rollershutter here?

File: bundles/org.openhab.binding.mqtt.generic/src/main/java/org/openhab/binding/mqtt/generic/values/OnOffValue.java [hunk 3]
```diff
@@ -44,55 +43,40 @@ public class OnOffValue extends Value {
...
         super(CoreItemFactory.SWITCH, List.of(OnOffType.class, StringType.class));
-        this.onStates = Stream.of(onStates).filter(not(String::isBlank)).collect(Collectors.toSet());
-        this.offStates = Stream.of(offStates).filter(not(String::isBlank)).collect(Collectors.toSet());
-        this.onCommand = onCommand;
-        this.offCommand = offCommand;
+        this.onStates = onStates == null ? Set.of(OnOffType.ON.name())
+                : onStates.stream().filter(not(String::isBlank)).collect(Collectors.toSet());
+        this.offStates = offStates == null ? Set.of(OnOffType.ON.name())
+                : offStates.stream().filter(not(String::isBlank)).collect(Collectors.toSet());
+        this.onCommand = onCommand == null ? (onStates == null ? OnOffType.ON.name() : onStates.get(0)) : onCommand;
```
Comment (jimtng) [inline bundles/org.openhab.binding.mqtt.generic/src/main/java/org/openhab/binding/mqtt/generic/values/OnOffValue.java:77]: I'm wondering whether onStates could be not null, but having size() == 0, in which case, get(0) will cause index out of bounds.

File: bundles/org.openhab.binding.mqtt.generic/src/main/resources/OH-INF/config/switch-channel-config.xml [hunk 4]
```diff
@@ -79,15 +79,25 @@
...
-			<label>Custom Off/Closed Value</label>
-			<description>A number (like 0, -10) or a string (like "disabled") that is additionally recognised as off/closed
-				state. You can use this parameter for a second keyword, next to OFF (CLOSED respectively on a Contact).</description>
+		<parameter name="offState" type="text" multiple="true">
+			<label>Custom Off/Closed Values</label>
+			<description>A list of numbers (like 0, -10) or strings (like "disabled") which are additionally recognised as
+				off/closed
+				state. You can use this parameter for additional keywords, next to OFF (CLOSED respectively on a Contact).</description>
+		</parameter>
+		<parameter name="on" type="text" advanced="true">
```
Comment (ccutrer) [inline bundles/org.openhab.binding.mqtt.generic/src/main/resources/OH-INF/config/switch-channel-config.xml:94]: why rename the parameter? I know it makes it somewhat clearer, but it also means people will need to update their configuration at some point. my understanding from @jimtng's investigations are that you can add `multiple="true"` to an existing parameter, and then the core framework will automatically handle always presenting it to the binding as a list, regardless of if it's actually a list in the persisted configuration (either json or .things file)

File: bundles/org.openhab.binding.mqtt.generic/src/main/java/org/openhab/binding/mqtt/generic/values/OnOffValue.java [hunk 5]
```diff
@@ -44,55 +43,40 @@ public class OnOffValue extends Value {
...
      *
-     * values send in messages will be the same as those expected in incoming messages
-     *
-     * @param onValue The ON value string. This will be compared to MQTT messages. Defaults to "ON".
-     * @param offValue The OFF value string. This will be compared to MQTT messages. Defaults to "OFF".
+     * @param onValues The list of ON value strings. These will be compared to MQTT messages. Defaults to "ON" if null.
+     * @param offValues The list of OFF value strings. These will be compared to MQTT messages. Defaults to "OFF" if
+     *            null.
      */
-    public OnOffValue(@Nullable String onValue, @Nullable String offValue) {
```
Comment (ccutrer) [inline bundles/org.openhab.binding.mqtt.generic/src/main/java/org/openhab/binding/mqtt/generic/values/OnOffValue.java:58]: maintaining this constructor as a convenience seems quite useful for uses _besides_ the mqtt.generic binding (i.e. from within homie and homeassistant, and tests in general)

File: bundles/org.openhab.binding.mqtt.generic/README.md [hunk 6]
```diff
@@ -128,13 +128,13 @@ You can connect this channel to a Rollershutter or Dimmer item.
 - **on**: An optional number (like 1, 10) or a string (like "ON"/"Open") that is recognized as on/open state.
 - **off**: An optional number (like 0, -10) or a string (like "OFF"/"Close") that is recognized as off/closed state.
 
+You can provide multiple values for **on** and **off** separated by a comma. The first value is used for publishing to MQTT.
```
Comment (jimtng) [inline bundles/org.openhab.binding.mqtt.generic/README.md:131]: this needs to be updated

File: bundles/org.openhab.binding.mqtt.generic/src/main/java/org/openhab/binding/mqtt/generic/values/OpenCloseValue.java (hunk 1 above)
Comment (florian-h05) [reply bundles/org.openhab.binding.mqtt.generic/src/main/java/org/openhab/binding/mqtt/generic/values/OpenCloseValue.java:37]: Normally it's not possible to send commands to Contact Items, so I guess it's unlikely a command is sent here, however I am not sure if this code is possibly needed to "publish" a contact Item state to MQTT.

File: bundles/org.openhab.binding.mqtt.generic/src/main/resources/OH-INF/config/switch-channel-config.xml (hunk 4 above)
Comment (florian-h05) [reply bundles/org.openhab.binding.mqtt.generic/src/main/resources/OH-INF/config/switch-channel-config.xml:94]: The problem is, that I would need to change the type of the `on` and `off` params in the `ChannelConfig` class from `String` to `List<String>`, which would require changes to several types, whilst changes to the `onState` and `offState` param require less changes.

File: bundles/org.openhab.binding.mqtt.generic/src/main/java/org/openhab/binding/mqtt/generic/values/ValueFactory.java (hunk 2 above)
Comment (florian-h05) [reply bundles/org.openhab.binding.mqtt.generic/src/main/java/org/openhab/binding/mqtt/generic/values/ValueFactory.java:87]: This is implemented inside `RollershutterValue`: It gets the `List<String>` onState and offState passed.

File: bundles/org.openhab.binding.mqtt.generic/src/main/java/org/openhab/binding/mqtt/generic/values/OnOffValue.java (hunk 3 above)
Comment (florian-h05) [reply bundles/org.openhab.binding.mqtt.generic/src/main/java/org/openhab/binding/mqtt/generic/values/OnOffValue.java:77]: Could be possible, true.

File: bundles/org.openhab.binding.mqtt.generic/src/main/resources/OH-INF/config/switch-channel-config.xml (hunk 4 above)
Comment (florian-h05) [reply bundles/org.openhab.binding.mqtt.generic/src/main/resources/OH-INF/config/switch-channel-config.xml:94]: This makes it really difficult IMO to implement the changes I want to do in a "clean" way - FYI I have already changed my approach and actually don't need the changes from this PR at all. Just use a SCRIPT transformation on the Item channel link and you can do whatever you want ...

Comment (jimtng) [reply bundles/org.openhab.binding.mqtt.generic/src/main/resources/OH-INF/config/switch-channel-config.xml:94]: > FYI I have already changed my approach and actually don't need the changes from this PR at all.

What's your plan for this PR? continue or drop ?

Comment (florian-h05) [reply bundles/org.openhab.binding.mqtt.generic/src/main/resources/OH-INF/config/switch-channel-config.xml:94]: I would tend to drop, as it is just way easier to evaluate the string with a script transformation ...
I will close this PR then.

**Pull request closed without merging**